### PR TITLE
Tutorial: Rust FFI in Haskell

### PR DESCRIPTION
**Pull request by shivaraj-bh**
preview:
```sh
nix run github:nixos-asia/website/haskell-rust-ffi#preview -- --port 5555
open http://localhost:5555/en/haskell-rust-ffi
```

### CODE REVIEW

**Comment by shivaraj-bh** [comment]
I wanted to write a blog post about this, rather than publishing it as a tutorial in either [community.flake.parts](https://community.flake.parts/haskell-flake) or here, as I wanted to describe the entire process, as I debugged, in the form of a story.

**Comment by srid** [comment]
Is this still being written?

You also want to,

- [ ] Move the post to the [`./blog`](https://github.com/nixos-asia/website/tree/master/en/blog) folder
- [ ] Check off #29

**Comment by shivaraj-bh** [comment]
> Is this still being written?

Yes, I have to resume work on this

**Comment by srid** [comment]
Associated repo (I think): https://github.com/shivaraj-bh/haskell-rust-ffi-template

**Comment by shivaraj-bh** [comment]
It is more appropriate as a tutorial series under [Nix for development](https://nixos.asia/en/tutorial), so I have made the necessary changes and added placeholders, will be filling them up soon.

File: en/haskell-rust-ffi.md
Comment (shivaraj-bh) [inline en/haskell-rust-ffi.md:133]: - [ ] Mention why we use `rust_nix_template` here while the rust package is called `rust-nix-template`

Comment (shivaraj-bh) [reply en/haskell-rust-ffi.md:126]: https://github.com/nixos-asia/website/pull/24/commits/00a2bf2a5266223da7b28d2f8abc938209cae454

File: en/haskell-rust-ffi.md
Comment (srid) [inline en/haskell-rust-ffi.md:185]:
```suggestion
> If you are on [[macos]], the Haskell package will not run because `dlopen` will be looking for the `.dylib` file in the temporary build directory (`/private/tmp/nix-build-rust-nix...`). To fix this, you need to include [fixDarwinDylibNames](https://github.com/NixOS/nixpkgs/blob/af8fd52e05c81eafcfd4fb9fe7d3553b61472712/pkgs/build-support/setup-hooks/fix-darwin-dylib-names.sh) in `flake.nix`:
```

File: en/haskell-rust-ffi.md
Comment (srid) [inline en/haskell-rust-ffi.md:184]:
```suggestion
> [!note] macOS caveat
```

**Review by srid — APPROVED** [review]
LGTM, after these changes.

File: en/haskell-rust-ffi.md
Comment (shivaraj-bh) [reply en/haskell-rust-ffi.md:59]: This will break the further steps in the tutorial, as we are assuming that the `.cabal` file generated by `cabal-install` in `flake:nixpkgs` will work with the `nixpkgs` input of the flake in the tutorial.

This was the initial command, but when I was following along the tutorial it fails, as constraints on `base` depends on `GHC` version, which can be different based on the nixpkgs commit. I doubt there is a way to `cabal init` without any constraints on the `base` package.

Comment (srid) [reply en/haskell-rust-ffi.md:59]: Pass `-d base` option?

```
❯ nix shell nixpkgs#ghc nixpkgs#cabal-install -c cabal -- init -n --exe -m --simple -d base --overwrite
Warning: The package list for 'hackage.haskell.org' does not exist. Run 'cabal
update' to download it.
[Log] Using cabal specification: 3.0
[Warning] unknown license type, you must put a copy in LICENSE yourself.
[Log] CHANGELOG.md already exists. Backing up old version in CHANGELOG.md.save0
[Log] Overwriting file CHANGELOG.md...
[Log] app already exists. Backing up old version in app.save1
[Log] Overwriting directory ./app...
[Log] Creating fresh file app/Main.hs...
[Log] hello-haskell.cabal already exists. Backing up old version in hello-haskell.cabal.save0
[Log] Overwriting file hello-haskell.cabal...
[Warning] No synopsis given. You should edit the .cabal file and add one.
[Info] You may want to edit the .cabal file and add a Description field.


srid on appreciate.intern /tmp/hello-haskell via λ
❯ rg base *.cabal
14:    build-depends:    base
```

Comment (shivaraj-bh) [reply en/haskell-rust-ffi.md:59]: Thanks!

Comment (shivaraj-bh) [reply en/haskell-rust-ffi.md]: https://github.com/nixos-asia/website/pull/24/commits/b28fabe260c971bfed4d7dea3aa84c8106321ffc

**Comment by srid** [comment]
@shivaraj-bh Feel free to announce in https://twitter.com/nixos_asia, Zulip, and elsewhere as you see fit.

**Comment by shivaraj-bh** [comment]
I probably should’ve linked this earlier here, but I was waiting on this before merging: https://github.com/srid/haskell-flake/issues/283#issuecomment-2031339304

**Comment by srid** [comment]
Okay let's announce it after I fix that issue.

**Pull request merged**
